content/docs/document-lifecycle.md

```diff
@@ -57,13 +63,15 @@ B --> D(<b>AstNodeLocator</b><br><br><i>gets the path of an AstNode</i>)
 {{</mermaid>}}
 
 ## Pre-processing
-This step is executed on 'LangiumDocument's with the state `IndexedContent` and regroups all necessary steps that needs to be done **prior to** resolving cross-references.
+This phase is executed on documents with the state `IndexedContent` and unites all necessary steps that need to be done **prior to** resolving cross-references.
 
-By default, the pre-processing consists of computing the scope of the AST. The default implementation of the `ScopeComputation` service computes the scope for **every** node in the AST and makes named nodes visible to their container. This means that the container holds information about which named nodes are nested inside of it. This is then used during linking to find the closest `AstNode` that matches a reference.
+By default, the pre-processing consists of gathering all symbols of the AST (both local and exported), done by the `ScopeComputation` service. Metadata of the gathered symbols are represented with `AstNodeDescription` like in the [initial indexing phase](#indexing-of-symbols). These metadata are attached to the `LangiumDocument` in a multi-map structure that associates a (possibly empty) set of symbol descriptions to each container node of the AST, called the *precomputed scopes*. These are used in the linking phase to construct the actual *scope* of a cross-reference, i.e. all possible symbols that are reachable. A symbol in the precomputed scopes is reachable from a specific cross-reference if it is associated with a direct or indirect container of that reference. Symbols associated to the root node are reachable from the whole AST, while symbols associated with an inner node are reachable from the respective sub-tree.
 
-After all pre-processing steps have been completed, the document's state is set to `Processed`.
+The default implementation of the `ScopeComputation` service attaches the description of every symbol to its direct container. This means that the container holds information about which named nodes are nested inside of it. You can override this default behavior to change the position where a symbol is reachable, or to change the name by which it can be referenced. It is even possible to associate the same symbol to multiple container nodes, possibly with different names, to control precisely where and how references to it can be resolved. However, keep in mind that you cannot access any cross-references in this phase. More complex, context-dependent scope mechanisms can be implemented in the `ScopeProvider` (see below).
```

Comment:
This might be difficult to follow without a concrete example (maybe we could link to the domainmodel example which implements a custom ScopeComputation?)